cli/flox-rust-sdk/src/models/environment/generations.rs

```diff
@@ -500,13 +471,41 @@ fn read_metadata(
     repo: &GitCommandProvider,
     ref_name: &str,
 ) -> Result<AllGenerationsMetadata, GenerationsError> {
-    let metadata = {
-        let metadata_content = repo
-            .show(&format!("{}:{}", ref_name, GENERATIONS_METADATA_FILE))
-            .map_err(GenerationsError::ShowMetadata)?;
-        serde_json::from_str(&metadata_content.to_string_lossy())
-            .map_err(GenerationsError::DeserializeMetadata)?
+    let metadata_content = repo
+        .show(&format!("{}:{}", ref_name, GENERATIONS_METADATA_FILE))
+        .map_err(GenerationsError::ShowMetadata)?;
+
+    parse_metadata(&mut serde_json::Deserializer::from_slice(
+        metadata_content.as_bytes(),
+    ))
+}
+
+fn parse_metadata<'de>(
+    deserializer: impl Deserializer<'de, Error = serde_json::Error>,
+) -> Result<AllGenerationsMetadata, GenerationsError> {
+    #[derive(Debug, Deserialize)]
+    #[serde(untagged)]
+    enum MetadataVersionCompat {
+        V1(compat::AllGenerationsMetadataV1),
+        V2(AllGenerationsMetadata),
+        VX { version: serde_json::Value },
+    }
+
+    let metadata: MetadataVersionCompat = MetadataVersionCompat::deserialize(deserializer)
+        .map_err(GenerationsError::DeserializeMetadata)?;
+
+    let metadata = match metadata {
+        MetadataVersionCompat::V1(all_generations_metadata_v1) => {
+            all_generations_metadata_v1.try_into()?
+        },
+        MetadataVersionCompat::V2(all_generations_metadata) => all_generations_metadata,
+        MetadataVersionCompat::VX { version } => Err(GenerationsError::DeserializeMetadata(
+            serde_json::Error::custom(format!(
+                "Environemnt metadata of version '{version}' is not supported",
```

Comment:
This will be triggered by bumping the metadata version when adding new `HistoryKind` variants in the future, so we should provide a hint that they should try upgrading their Flox version. Haven't left a code suggestion though because I'm not sure if we want to emit multi-line errors from this part of the SDK?

The other possible but unlikely cause is that the metadata file has been manually edited or corrupted.

Reply:
> record the actual command invoked 

are you looking for the "subcommand" or the whole `argv`?

> With this in mind, having a strongly-typed "kind" doesn't just introduce compatibility issues but would also tend to be redundant when we make a record of the original command invoked.

As you say multiple commands can have side effects, like upgrading and i do think its reasonable to filter for "upgrades", regardless of which command that was.